plasmapy/diagnostics/thomson.py

```diff
@@ -204,114 +444,372 @@ def spectral_density(
+
+    # Automatically add an expression to the last efract parameter to
+    # indicate that it depends on the others (so they sum to 1.0)
+    # The resulting expression for the last of three will look like
+    # efract_2.expr = "1.0 - efract_0 - efract_1"
+    if num_e > 1:
+        nums = ["efract_" + str(i) for i in range(num_e - 1)]
+        nums.insert(0, "1.0")
+        params["efract_" + str(num_e - 1)].expr = " - ".join(nums)
+
+    if num_i > 1:
+        nums = ["ifract_" + str(i) for i in range(num_i - 1)]
+        nums.insert(0, "1.0")
+        params["ifract_" + str(num_i - 1)].expr = " - ".join(nums)
+
+    # **************
+    # Electron velocity
+    # **************
+    electron_speed = np.zeros([num_e])
+    for e in range(num_e):
+        k = "electron_speed_" + str(e)
+        if k in list(params.keys()):
+            electron_speed[e] = params[k].value
+        else:
+            # electron_speed[e] = 0 already
+            params.add(k, value=0, vary=False)
+
+    if "electron_vdir" not in list(settings.keys()):
+        if np.all(electron_speed == 0):
+            # vdir is arbitrary in this case because vel is zero
+            settings["electron_vdir"] = np.ones([num_e, 3])
+        else:
+            raise ValueError(
+                "electron_vdir must be set if electron_speeds " "are not all zero."
+            )
+    # Normalize vdir
+    norm = np.linalg.norm(settings["electron_vdir"], axis=-1)
+    settings["electron_vdir"] = settings["electron_vdir"] / norm[:, np.newaxis]
+
+    # **************
+    # Ion velocity
+    # **************
+    ion_speed = np.zeros([num_i])
+    for i in range(num_i):
+        k = "ion_speed_" + str(i)
+        if k in list(params.keys()):
+            ion_speed[i] = params[k].value
+        else:
+            # ion_speed[i] = 0 already
+            params.add(k, value=0, vary=False)
+
+    if "ion_vdir" not in list(settings.keys()):
+        if np.all(ion_speed == 0):
+            # vdir is arbitrary in this case because vel is zero
+            settings["ion_vdir"] = np.ones([num_i, 3])
+        else:
+            raise ValueError("ion_vdir must be set if ion_speeds " "are not all zero.")
+    # Normalize vdir
+    norm = np.linalg.norm(settings["ion_vdir"], axis=-1)
+    settings["ion_vdir"] = settings["ion_vdir"] / norm[:, np.newaxis]
+
+    if "inst_fcn" not in list(settings.keys()):
+        settings["inst_fcn_arr"] = None
+    else:
+        # Create inst fcn array from inst_fcn
+        inst_fcn = settings["inst_fcn"]
+        wspan = (np.max(wavelengths) - np.min(wavelengths)) / 2
+        eval_w = np.linspace(-wspan, wspan, num=wavelengths.size)
+        inst_fcn_arr = inst_fcn(eval_w * u.m)
+        inst_fcn_arr *= 1 / np.sum(inst_fcn_arr)
+        settings["inst_fcn_arr"] = inst_fcn_arr
+
+    # TODO: raise an exception if the number of any of the ion or electron
+    # quantities isn't consistent with the number of that species defined
+    # by ifract or efract.
```

Comment:
[optional] IDEs tend to recognize TODO better when the comment is indented to the same level

```suggestion
    # TODO: raise an exception if the number of any of the ion or electron
    #       quantities isn't consistent with the number of that species defined
    #       by ifract or efract.
```